demos/instr.h

```diff
@@ -182,38 +177,76 @@ inline void BoundsCheck(const char *str, size_t offset) {
   asm volatile(".byte 0x62, 0x02"::"a"(offset), "d"(&string_bounds):"memory");
 }
 
-// Reads an offset from the FS segment.
+// Reads an offset from the SS segment.
 // Must be inlined because the fault occurs inside and the stack pointer would
 // be shifted.
 // We fetch offset + 1 from the segment base, because the base is shifted one
 // byte below to bypass 1-byte minimal segment size.
 SAFESIDE_ALWAYS_INLINE
-inline unsigned int ReadUsingFS(unsigned int offset) {
+inline unsigned int ReadUsingSS(unsigned int ss, unsigned int offset) {
   unsigned int result;
 
   asm volatile(
-      "movzbl %%fs:(, %1, 1), %0\n"
-      :"=r"(result):"r"(offset + 1):"memory");
+      "movl %1, %%ss\n"
+      // Open a possibility for a speculation window.
+      ".rept 1000\n"
+      "xorl %1, %1\n"
+      ".endr\n"
+      "movzbl %%ss:(, %2, 1), %0\n"
+      :"=d"(result):"a"(ss),"b"(offset + 1):"memory");
 
   return result;
 }
 
-// Reads an offset from the ES segment.
+// Returns the original value of SS.
+SAFESIDE_ALWAYS_INLINE
+inline unsigned int BackupSS() {
```

Comment:
Slight preference for `GetSS` and `SetSS` since "Backup" and "Restore" have more to do with how they're used.